### PR TITLE
improve logging

### PR DESCRIPTION
**Pull request by TornaxO7**
# Description

Replacing `slog` with `env_logger`.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Updated user documentation:

Not needed.

# Checklist:

- [x] Ran `make test-full` locally with no errors or warnings reported
- [x] Manual page has been updated accordingly
- [x] Wiki pages have been updated accordingly (to perform **after** merge)

### CODE REVIEW

**Comment by TornaxO7** [comment]
Hm... I don't receive any logs yet, I need some time to investigate.

**Comment by TornaxO7** [comment]
Alright, it should work now!

File: leftwm/src/bin/leftwm-worker.rs
Comment (TornaxO7) [inline leftwm/src/bin/leftwm-worker.rs:36]: Is it fine, if we use `debug` as the defualt log-level?

File: leftwm/src/utils/cacher.rs
Comment (VuiMuich) [inline leftwm/src/utils/cacher.rs:4]: Is there possibly a way to avoid introducing this new dependency?

**Review by VuiMuich — COMMENTED** [review]
I would like to keep the possibility to log to `journald`. For developing logging to stdout or stderr is fine, but the whole idea of #508 was to make logging to journald the default for packages like in the AUR so troubleshooting for 'non dev users' gets another valuable tool.
Maybe adding `libsystemd` under the `journald` feature flag might be good, but I have no idea how big the effort is to add the target: https://docs.rs/libsystemd/0.4.1/libsystemd/logging/index.html

Do you see any further use for the `cacher` despite logging? I would suggest to refactor anything related to a `leftwm/src/logging.rs` module. And IMO that would be the place for the `cacher` if no other place might use it.

And finally, I don't want to generally block now dependencies, but I would like to have at least a little discussion about the necessity, to have some references for devs in the future.

**Comment by TornaxO7** [comment]
> Do you see any further use for the cacher despite logging?

Yes I do! Currently we are using `/tmp` for the down script for example or the state, which isn't that good in my opinion, because everyone can read it which might be unwanted for some users.

File: leftwm/src/utils/cacher.rs
Comment (TornaxO7) [reply leftwm/src/utils/cacher.rs:4]: Hm... I thought that having a singleton would be the best since it would be kinda weird to have multiple "cachers". So the answer is actually no sadly.... but I'll ask my friends if they've got any ideas.

**Comment by VuiMuich** [comment]
> > Do you see any further use for the cacher despite logging?
> 
> Yes I do! Currently we are using `/tmp` for the down script for example or the state, which isn't that good in my opinion, because everyone can read it which might be unwanted for some users.

Valid point. :+1:

**Comment by TornaxO7** [comment]
> Plus how would you access this cacher from a script?

Oh right.... I had in my head, that `leftwm` takes care of the themes.... damn...

**Comment by VuiMuich** [comment]
> Hm... I understand this point but couldn't it be a disadvantage for non-systemd users? I'll rethink the logging system a bit

Absolutely, so the `journald` logging should not necessary be exclusive of logging to stdout or a logfile.
Ideally there would be a feature-flag that builds with `journald` and IMO this should be included in `default`.
Packaging scripts might check if `systemd` is being used and if not (or in case of the `void` package) build with `--no-default-features`.

File: leftwm/src/bin/leftwm-worker.rs
Comment (VuiMuich) [inline leftwm/src/bin/leftwm-worker.rs:13]: I guess this was inadvertedly reverted, right?
```suggestion
    let exit_status = panic::catch_unwind(|| {
```

File: leftwm/src/bin/leftwm-worker.rs
Comment (VuiMuich) [reply leftwm/src/bin/leftwm-worker.rs]: I would be ok with setting the debug level in `dev` builds and info in `release` ones

**Comment by AethanFoot** [comment]
For logging to journald (I am not 100% if this works), but could we use syslog. This then in theory would log to any syslog daemon, e.g. for void https://docs.voidlinux.org/config/services/logging.html .

**Comment by TornaxO7** [comment]
> For logging to journald (I am not 100% if this works), but could we use syslog.

Syslog looks pretty promising for unix-based systems but it looks like that it doesn't create an opportunity for journald according to [this list](https://docs.rs/syslog/6.0.1/syslog/fn.unix.html). We could use the [`unix`](https://docs.rs/syslog/6.0.1/syslog/fn.unix.html) function which uses the `/var/run/log`.

**Comment by TornaxO7** [comment]
Hm... or should be keep using `slog` but with a nice little wrapper to simplify the usage with it?

**Comment by TornaxO7** [comment]
@VuiMuich @AethanFoot

**Comment by AethanFoot** [comment]
I think journald picks up logs from those files (if you use `logger hello` it prints to journalctl, and logger uses syslog, same with `python -c 'import syslog; syslog.syslog("Hello World")'`). Yeah we could create a simpler use, as slog has an env_logger and syslog extensions.

**Comment by AethanFoot** [comment]
We do need to overhaul error handling and messages at some point as well, which is related to this (if we make it its own crate at some point, lefthk can use it as well)

**Comment by TornaxO7** [comment]
> Yeah we could create a simpler use, as slog has an env_logger and syslog extensions.

What do you mean with this? So it's fine if we keep slog?

**Comment by AethanFoot** [comment]
Yeah we could, or we could the env_logger with a syslog crate. Either way would be okay.

**Comment by TornaxO7** [comment]
@AethanFoot @VuiMuich A general question how the logging should work: Should the user be able to log to multiple places simultaneously like for example to journald + a log file at the same time or should we only enable logging only for *one* "type"?

**Comment by AethanFoot** [comment]
I feel that for most users logging to journald (or their relevant syslog thing) should be default. Then logging to file can be enable for someone doi g dev work and disabling logging through features for users minimising.

**Comment by TornaxO7** [comment]
@AethanFoot yes, but what's with logging to multiple places? Do think that's needed? Like logging to journald + to a file?

**Comment by AethanFoot** [comment]
I would say yes, however we shouldnt need to worry too much if the two options are setup separately for their purpose. If they enable logging to file then yes both should be enabled. If they do nothing just syslog. And they can disable completely. (Sorry just trying to layout the cases clearly)

**Comment by TornaxO7** [comment]
It would be nice if someone could test the logging with `syslog` and `journald` because I'm using `artix` without systemd (and so journald). I don't really understand how to wrtie to `/var/log/syslog.log`, I even tried to write with C into it:
```c
#include <syslog.h>

int main() {
    openlog("Artix btw", LOG_PID, LOG_USER);
    syslog(LOG_INFO, "YEEEEEEEEEE");
    closelog();
}
```
but that didn't work either.

The file works fine at my last test.

**Comment by VuiMuich** [comment]
Just a quick feedback, a more in-depth review might follow this evening:
- Personally I'd prefer the inline imports over the imports over the whole file for mainly the reason, that with the inline imports it is easy to rip-grep for all "logging" calls no matter their level in one go
- I merged this over the `lefthk` branch, the merge conflicts are easy to solve and it runs well so far. It logs to journald, by default only at level info, which is fine but I don't see yet how to raise verbosity. Maybe we should besides later documenting this in the wiki leave a doc comment at the setup-logging invocation in `leftwm-worker` how to do this

Besides that great work and awesome find with the `tracing` crate, iirc the previous crate count we are now down by ten crates in total.

File: leftwm/src/bin/leftwm-worker.rs
Comment (VuiMuich) [inline leftwm/src/bin/leftwm-worker.rs]: I loved how you could see the struggle that went into getting there :heart_eyes:

**Comment by TornaxO7** [comment]
> Personally I'd prefer the inline imports over the imports over the whole file for mainly the reason, that with the inline imports it is easy to rip-grep for all "logging" calls no matter their level in one go

So should I change it back @VuiMuich ?

**Comment by VuiMuich** [comment]
@0323pin my I invite you to test this branch built with `cargo b --release --no-default-features -features=lefthk,sys-log`
Also would you see any issues with bumping msrv to `1.57.0`?

**Comment by VuiMuich** [comment]
this closes #508

**Comment by 0323pin** [comment]
@VuiMuich Sure, no worries.
Ok, if I build this tomorrow?

**Comment by 0323pin** [comment]
@VuiMuich 
> Also would you see any issues with bumping msrv to 1.57.0?

Nope, NetBSD is on 1.62.1 and I'm on 1.63.0

**Comment by VuiMuich** [comment]
> @VuiMuich Sure, no worries. Ok, if I build this tomorrow?

Sure



> @VuiMuich
> 
> > Also would you see any issues with bumping msrv to 1.57.0?
> 
> Nope, NetBSD is on 1.62.1 and I'm on 1.63.0

Nice. I think we still bump to the minimum we need to suffice our deps.

**Comment by 0323pin** [comment]
Ok! I'll give this a go tomorrow.

I hope none of you guys take this wrong but, ...
please, **no** systemd and **no** polkit. I'm fine with dbus but, that's as far as I can go.

**Comment by VuiMuich** [comment]
> Ok! I'll give this a go tomorrow.
> 
> 
> 
> I hope none of you guys take this wrong but, ...
> 
> please, **no** systemd and **no** polkit. I'm fine with dbus but, that's as far as I can go.

Of course we don't intend to only support systemd, we just want it as default for systems running on systemd to be able to better troubleshoot user issues.

For users on non-systemd systems the 'default' should be syslog and for development we keep logging to a file around as an option.

Right now we do not automatically default to syslog if systemd is not available. AFAIK cargo is not able to do such checks and logic, but the Makefile should and package maintainers ofc can set those flags for their systems packages.

**Comment by 0323pin** [comment]
@VuiMuich Just built from 3bbbc94843294f4af15c2a5ee51eb973e57d4e76, all good.
Just need to convert my config to `ron`

> package maintainers ofc can set those flags for their systems packages

Yes, I need to look into this though.

**Comment by 0323pin** [comment]
@VuiMuich Oh! How do I assign the old layout_mode to Tag instead of workspace in the new ron config file?

**Comment by VuiMuich** [comment]
I have this (giving layouts and scratchpad for context):
```rust
layouts: [
    MainAndDeck,
    Monocle,
    CenterMainFluid
],
layout_model: Tag,

scratchpad: <...>
```

**Comment by 0323pin** [comment]
@VuiMuich Yeah, I found it.

What about theme.toml? Should that also be converted to ron?
I have a very simple theme.toml file but, how to convert this to ron?

```
# leftwm theme.toml

border_width          = 0
margin                = 4
default_border_color  = "#282828"
floating_border_color = "#C0C5CE"
focused_border_color  = "#C0C5CE"
```

**Comment by 0323pin** [comment]
Fixed 😄 

```
(
    border_width: 0,
    margin: 4,
    default_border_color: "#282828",
    floating_border_color: "#C0C5CE",
    focused_border_color: "#C0C5CE",
)
```

**Comment by VuiMuich** [comment]
> border_width          = 0
> margin                = 4
> default_border_color  = "#282828"
> floating_border_color = "#C0C5CE"
> focused_border_color  = "#C0C5CE"

This should do:
```rust
#![enable(implicit_some)]
(
border_width:          0,
margin:                4,
default_border_color:  "#282828",
floating_border_color: "#C0C5CE",
focused_border_color:  "#C0C5CE",
)
```

**Comment by 0323pin** [comment]
> I guess a migration guide for themes should be added to the wiki as well.

Indeed.
But, I got it working anyway.

**Comment by 0323pin** [comment]
> I guess the mention of polket was just a precaution, as I don't remember any discussion about it so far, right?

Yes, it was just a precaution.

**Comment by 0323pin** [comment]
@VuiMuich and @TornaxO7 Feel free to merge this whenever you guys think it's ready.
I have made up my mind on _how  to_ switch-off `journald-log`.

**Comment by TornaxO7** [comment]
> I have made up my mind on how to switch-off journald-log.

@0323pin so you understood how to change it?

**Comment by 0323pin** [comment]
@TornaxO7 Yes, I'm running a build from your branch right now, no worries.
I maintain about 80 rust packages on NetBSD, so I need to find my way around things like this 😄 

Whenever you guys are ready, I'm ready.

**Comment by TornaxO7** [comment]
@VuiMuich @AethanFoot @hertg Review can start I think (I guess only 2 reviews are fine as well?)

**Comment by TornaxO7** [comment]
@0323pin you can review it as well, if you want to of course :)

**Comment by 0323pin** [comment]
@TornaxO7 Thanks for the offer :)

> Right now we do not automatically default to syslog if systemd is not available. AFAIK cargo is not able to do such checks and logic, but the Makefile should and package maintainers ofc can set those flags for their systems packages.

My first though was to use `[target.'cfg(target_os = "linux")'.dependencies]` and `[target.'cfg(all(unix, not(target_os = "linux")))'.dependencies]` in `Cargo.toml`. But, this is ofc not possible as it would exclude all non-systemd linux systems.

So, I will just do,

```
CARGO_NO_DEFAULT_FEATURES=	YES
CARGO_FEATURES+=		lefthk sys-log
```

If someone is using pkgsrc on a systemd-linux system, ... well, bad luck.

**Comment by VuiMuich** [comment]
> My first though was to use `[target.'cfg(target_os = "linux")'.dependencies]` and `[target.'cfg(all(unix, not(target_os = "linux")))'.dependencies]` in `Cargo.toml`. But, this is ofc not possible as it would exclude all non-systemd linux systems.
> 
> 
> 
> So, I will just do,
> 
> 
> 
> ```
> 
> CARGO_NO_DEFAULT_FEATURES=	YES
> 
> CARGO_FEATURES+=		lefthk sys-log
> 
> ```

Fair enough. 

> If someone is using pkgsrc on a systemd-linux system, ... well, bad luck.

Well in that case that's just another quest in leftwm-adventure-land then 😁

**Comment by TornaxO7** [comment]
Well, why aren't we using [`build.rs`](https://doc.rust-lang.org/cargo/reference/build-scripts.html) then? We could simply use rusts `cfg(windows)`, `cfg(unix)` etc. to set the feature-flags.

**Comment by 0323pin** [comment]
> Well, why aren't we using [build.rs](https://doc.rust-lang.org/cargo/reference/build-scripts.html) then? We could simply use rusts cfg(windows), cfg(unix) etc. to set the feature-flags.

Hmm ... https://doc.rust-lang.org/reference/conditional-compilation.html says:
```
target_os

Example values:

"windows"
"macos"
"ios"
"linux"
"android"
"freebsd"
"dragonfly"
"openbsd"
"netbsd"
```

How would you define Linux with systemd and Linux without systemd?

**Comment by 0323pin** [comment]
Ok, been running a build from 5fcb5f6a5d92684c49175be81f793c39fad1945f for the last two days, no issues found.

**Comment by VuiMuich** [comment]
> Ok, been running a build from [5fcb5f6](https://github.com/leftwm/leftwm/commit/5fcb5f6a5d92684c49175be81f793c39fad1945f) for the last two days, no issues found.

Great. Thanks for testing!

**Comment by TornaxO7** [comment]
May I ask what we're waiting for? :eyes:

**Comment by VuiMuich** [comment]
We thought we'd merge #768 first and solve the resulting conflicts here, rather then the other way round and we almost merged the other two days ago, but then a few regressions put that on hold.

If you want to proceed to the next topic, don't feel obligated to wait for the merge of this here. The expected conflicts are not expected to be something major, we just wanted to spare the other contributor the chores of another 'merge main', as their branch had to undergo this already a couple of times.

**Comment by TornaxO7** [comment]
Understandable! Thank you for the information!

**Comment by VuiMuich** [comment]
btw:
Closes #508

**Review by VuiMuich — APPROVED** [review]

**Pull request merged**
